lib/spack/spack/database.py

```diff
@@ -339,7 +361,161 @@ def read_transaction(self):
         """Get a read lock context manager for use in a `with` block."""
         return ReadTransaction(self.lock, acquire=self._read)
 
-    def prefix_lock(self, spec):
+    def _failed_spec_path(self, spec):
+        """Return the path to the spec's failure file, which may not exist."""
+        if not spec.concrete:
+            raise ValueError('Concrete spec required for failure path for {0}'
+                             .format(spec.name))
+
+        return os.path.join(self._failure_dir,
+                            '{0}-{1}'.format(spec.name, spec.full_hash()))
+
+    def clear_failure(self, spec, force=False):
+        """
+        Remove any persistent and cached failure tracking for the spec.
+
+        Coordination between concurrent builds depends in large part on
+        prefix failure locks so the default is to not perform this operation
```

Comment:
"depends in large part" is vague -- say that failure files are created to prevent repeated retries of an unchanged package recipe within the same parallel spack invocation, and build processes hold locks on these files to prevent them from being removed while the build that generated the initial failure is still running.  Maybe mention that this is separate from the prefix locks -- those ensure that no two processes try to build the same thing at the same time.

Edit: or just say "see `mark_failed()`" for details (it's not clear where I get the big picture from the per-function comments here, which is another reason a separate class might be nice)

Reply:
Okay.  I saved off my modified description in case it comes in handy. :)